Make library compatible with `composer/composer:^2.0.14`, do not call `InstalledVersions::getRawData()` (which introduces runtime side-effect in pure API)

derrabus opened the pull request:
Follows composer/composer#9816

This PR proposes to not call `InstalledVersions::getRawData()` anymore. This method has been deprecated and triggers runtime deprecation warnings since Composer 2.0.14.

derrabus committed:
Don't call deprecated InstalledVersions::getRawData()

Ocramius commented:
Is there a way to put a constraint on `"composer-runtime-api": "^2.0.14"`
instead?

On Fri, May 21, 2021, 22:40 Alexander M. Turek ***@***.***>
wrote:

> Follows composer/composer#9816
> <https://github.com/composer/composer/pull/9816>
>
> This PR proposes to not call InstalledVersions::getRawData() anymore.
> This method has been deprecated and triggers runtime deprecation warnings
> since Composer 2.0.14.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Ocramius/ProxyManager/pull/700
> Commit Summary
>
>    - Don't call deprecated InstalledVersions::getRawData()
>
> File Changes
>
>    - *M* src/ProxyManager/Generator/Util/IdentifierSuffixer.php
>    <https://github.com/Ocramius/ProxyManager/pull/700/files#diff-f7d2e116d991592348fe79db9c4f41d46bd7593e78201834677cbc36097a4451>
>    (7)
>    - *M* tests/ProxyManagerTest/Generator/Util/IdentifierSuffixerTest.php
>    <https://github.com/Ocramius/ProxyManager/pull/700/files#diff-dee42a03aa90551faae473522a08d8cd938b67cdad1eecee2440fd91deaec03a>
>    (7)
>
> Patch Links:
>
>    - https://github.com/Ocramius/ProxyManager/pull/700.patch
>    - https://github.com/Ocramius/ProxyManager/pull/700.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Ocramius/ProxyManager/pull/700>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABFVEGLPZDVTHLK5BJAX3LTO3AL7ANCNFSM45J2P5VQ>
> .
>

derrabus commented:
I tried to do that, but apparently Composer 2.0.14 still claims to provide `"composer-runtime-api": "2.0.0"`.

derrabus commented:
```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires composer-runtime-api ^2.0.14, found composer-runtime-api[2.0.0] but it does not match the constraint.
```

Ocramius commented:
I wonder if we should conflict with `composer/composer:<2.0.14` then 🤔

Overall, feels weird that the plugin API doesn't reflect composer version:
we should probably report that in upstream.

Also, runtime deprecation warnings are reeeeeeally a no-go 😩

On Fri, May 21, 2021, 22:58 Alexander M. Turek ***@***.***>
wrote:

> Your requirements could not be resolved to an installable set of packages.
>
>   Problem 1
>     - Root composer.json requires composer-runtime-api ^2.0.14, found composer-runtime-api[2.0.0] but it does not match the constraint.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Ocramius/ProxyManager/pull/700#issuecomment-846256343>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABFVEB4RVJRQFFU3HGIUWTTO3CQZANCNFSM45J2P5VQ>
> .
>

derrabus commented:
cc @Seldaek

Seldaek commented:
I was pretty sure nobody relied on this method but of course @Ocramius to the rescue :p

Anyway yes should have bumped the runtime api on this one sorry. 2.1 may come out soon if i manage tho, so we'll see what makes it first..

That said IMO merging this as is would be better as it ensures broader compatibility and people are notoriously bad at keeping composer up to date..

Ocramius commented on src/ProxyManager/Generator/Util/IdentifierSuffixer.php:
```diff
@@ -48,6 +49,10 @@ public static function getIdentifier(string $name): string
-        return sha1(serialize(InstalledVersions::getRawData()));
+        return sha1(serialize(
+            is_callable([InstalledVersions::class, 'getAllRawData'])
+                ? InstalledVersions::getAllRawData() // Composer >= 2.0.14
```
Note: I'll add a `"conflict": {"composer/composer": "<2.0.14"}` before merging, and remove the conditional.

Ocramius replied:
b5ee473

Ocramius committed:
Dropped compatibility with `composer/composer:<2.0.14`

As discussed in https://github.com/Ocramius/ProxyManager/pull/700
As noted in https://github.com/Ocramius/ProxyManager/pull/700#discussion_r638675531

Since `composer/composer` broke BC in 2.0.14 by adding a runtime side-effect to a
functionally pure API (`InstalledVersions::getRawData()`) it is safest to move away
from it.

Since `composer-runtime-api:2.1.0` is not yet available, the safest approach is to
introduce a conflict with `composer/composer:<2.0.14` for now.

In the next release, we'll drop this `conflict` block and add bump `composer-runtime-api`
instead ( https://github.com/Ocramius/ProxyManager/issues/702 ).

Ocramius commented:
Thanks @derrabus! Rolling a release nao :-)

Tobion commented:
This breaks every app using composer <2.0.14 or am I missing something?!

> bin/console c:c
Symfony\Component\ErrorHandler\Error\UndefinedMethodError^ {#13173
  #message: """
    Attempted to call an undefined method named "getAllRawData" of class "Composer\InstalledVersions".\n
    Did you mean to call "getRawData"?
    """
  #code: 0
  #file: "./vendor/ocramius/proxy-manager/src/ProxyManager/Generator/Util/IdentifierSuffixer.php"
  #line: 51
  trace: {
    ./vendor/ocramius/proxy-manager/src/ProxyManager/Generator/Util/IdentifierSuffixer.php:

Ocramius commented:
If that's the case, then `composer/composer` is not declaring itself at runtime, which is potentially yet another upstream bug/problem :shrug:

Seldaek commented:
composer/composer is not available as a platform package no.. composer-runtime-api and composer-plugin-api are. Otherwise it wouldn't really be possible to add composer/composer as a dependency and have it installed in the vendor dir. Not something we're going to fix.

Ocramius commented:
Noted: for now, I'll point people observing crashes to this discussion then.

I've raised #702 for when the new plugin API is available.

Tobion commented on src/ProxyManager/Generator/Util/IdentifierSuffixer.php:
```diff
@@ -48,6 +48,6 @@ public static function getIdentifier(string $name): string
     private static function loadBaseHashSalt(): string
     {
-        return sha1(serialize(InstalledVersions::getRawData()));
+        return sha1(serialize(InstalledVersions::getAllRawData()));
```
Can't you add the `is_callable([InstalledVersions::class, 'getAllRawData'])` check here as well? This way, people won't run into this problem like I did.

Ocramius replied:
https://github.com/Ocramius/ProxyManager/pull/700#discussion_r638675531

Tobion replied:
yeah but it does not work

Venorcis commented:
> This breaks every app using composer <2.0.14 or am I missing something?!
> 
> > bin/console c:c
> > Symfony\Component\ErrorHandler\Error\UndefinedMethodError^ {#13173
> > #message: """
> > Attempted to call an undefined method named "getAllRawData" of class "Composer\InstalledVersions".\n
> > Did you mean to call "getRawData"?
> > """
> > #code: 0
> > #file: "./vendor/ocramius/proxy-manager/src/ProxyManager/Generator/Util/IdentifierSuffixer.php"
> > #line: 51
> > trace: {
> > ./vendor/ocramius/proxy-manager/src/ProxyManager/Generator/Util/IdentifierSuffixer.php:

I have composer 2.0.8 in my Docker instance and indeed this error pops up with 2.12.0
I'd suggest reverting the latest commit in master (keep the check on getRawData/getAllRawData)
